File: server-kotlin/src/main/kotlin/eu/lubsen/rummikub/idl/client/ClientMessages.kt
package eu.lubsen.rummikub.idl.client

import eu.lubsen.rummikub.core.parseMove
import eu.lubsen.rummikub.model.Lounge
import eu.lubsen.rummikub.model.Move
import io.vertx.core.json.JsonObject
import java.util.*

enum class ClientMessageType {
    CreateGame,
    RemoveGame,
    JoinGame,
    LeaveGame,
    StartGame,
    StopGame,
    RequestGameList,
    RequestPlayerList,
    RequestPlayerListForGame,
    RequestGameState,
    PlayerMove
}

sealed class ClientMessage(json: JsonObject) {
    abstract val type : ClientMessageType
    val playerId : UUID = UUID.fromString(json.getString(PLAYER_ID_FIELD))
}

class CreateGame(json: JsonObject) : ClientMessage(json = json) {
    val gameName : String = json.getString(GAME_ID_FIELD)
    override val type: ClientMessageType = ClientMessageType.CreateGame
}

class RemoveGame(json: JsonObject) : ClientMessage(json = json) {
    val gameName : String = json.getString(GAME_ID_FIELD)
    override val type: ClientMessageType = ClientMessageType.RemoveGame
}

class JoinGame(json: JsonObject) : ClientMessage(json = json) {
    val gameName : String = json.getString(GAME_ID_FIELD)
    override val type: ClientMessageType = ClientMessageType.JoinGame
}

class LeaveGame(json: JsonObject) : ClientMessage(json = json) {
    val gameName : String = json.getString(GAME_ID_FIELD)
    override val type: ClientMessageType = ClientMessageType.LeaveGame
}

class RequestGameList(json: JsonObject) : ClientMessage(json = json) {
    override val type: ClientMessageType = ClientMessageType.RequestGameList
}

class RequestPlayerList(json: JsonObject) : ClientMessage(json = json) {
    override val type: ClientMessageType = ClientMessageType.RequestPlayerList
}

class RequestPlayerListForGame(json: JsonObject) : ClientMessage(json = json) {
    override val type: ClientMessageType = ClientMessageType.RequestPlayerListForGame
    val gameName : String = json.getString(GAME_ID_FIELD)
}

class RequestGameState(json: JsonObject) : ClientMessage(json = json) {
    val gameName : String = json.getString(GAME_ID_FIELD);
    override val type: ClientMessageType = ClientMessageType.RequestGameState
}

class StartGame(json: JsonObject) : ClientMessage(json = json) {
    val gameName : String = json.getString(GAME_ID_FIELD)
    override val type: ClientMessageType = ClientMessageType.StartGame
}

class StopGame(json: JsonObject) : ClientMessage(json = json) {
    val gameName : String = json.getString(GAME_ID_FIELD)
    override val type: ClientMessageType = ClientMessageType.StopGame
}

class PlayerMove(json: JsonObject, lounge : Lounge) : ClientMessage(json = json) {
    val gameName : String = json.getString(GAME_ID_FIELD)
    val move : Move = parseMove(lounge, this, json)
    override val type: ClientMessageType = ClientMessageType.PlayerMove
}

const val GAME_ID_FIELD = "gameId"
const val PLAYER_ID_FIELD = "playerId"
const val MOVE_TYPE_FIELD = "moveType"
const val TILESET_ID_FIELD = "tileSetId"
const val SPLIT_SET_ID = "splitSetId"
const val SOURCE_ID_FIELD = "sourceId"
const val TARGET_ID_FIELD = "targetId"
const val SOURCE_LOCATION_FIELD = "sourceLocation"
const val TARGET_LOCATION_FIELD = "targetLocation"
const val MOVE_LOCATION_FIELD = "moveLocation"
const val INDEX_FIELD = "index"

File: client-svelte/src/idl/ClientMessages.js

function msgCreateGame(playerId, gameId) {
    return {
        messageType : "CreateGame",
        playerId : playerId,
        gameId : gameId
    }
}

function msgRemoveGame(playerId, gameId) {
    return {
        messageType : "RemoveGame",
        playerId : playerId,
        gameId : gameId
    }
}

function msgJoinGame(playerId, gameId) {
    return {
        messageType : "JoinGame",
        playerId : playerId,
        gameId : gameId
    }
}

function msgLeaveGame(playerId, gameId) {
    return {
        messageType : "LeaveGame",
        playerId : playerId,
        gameId : gameId
    }
}

function msgStartGame(playerId, gameId) {
    return {
        messageType : "StartGame",
        playerId : playerId,
        gameId : gameId
    }
}

function msgStopGame(playerId, gameId) {
    return {
        messageType : "StopGame",
        playerId : playerId,
        gameId : gameId
    }
}

function msgRequestPlayerList(playerId) {
    return {
        messageType : "RequestPlayerList",
        playerId : playerId
    }
}

function msgRequestPlayerListForGame(playerId, gameId) {
    return {
        messageType : "RequestPlayerListForGame",
        playerId : playerId,
        gameId : gameId
    }
}

function msgRequestGameList(playerId) {
    return {
        messageType : "RequestGameList",
        playerId : playerId
    }
}

function msgRequestGameState(playerId, gameId) {
    return {
        messageType : "RequestGameState",
        playerId : playerId,
        gameId : gameId
    }
}

function msgMerge(playerId, gameId, sourceId, targetId, index, location) {
    return {
        messageType : "PlayerMove",
        moveType : "MERGE",
        playerId : playerId,
        gameId : gameId,
        sourceId : sourceId,
        targetId : targetId,
        index : index,
        targetLocation : location
    }
}

function msgMoveAndMerge(playerId, gameId, sourceId, targetId, index, sourceLocation, targetLocation) {
    return {
        messageType : "PlayerMove",
        moveType : "MOVE_AND_MERGE",
        playerId : playerId,
        gameId : gameId,
        sourceId : sourceId,
        targetId : targetId,
        index : index,
        sourceLocation : sourceLocation,
        targetLocation : targetLocation
    }
}

function msgSplit(playerId, gameId, tileSetId, index, location) {
    return {
        messageType : "PlayerMove",
        moveType : "SPLIT",
        playerId : playerId,
        gameId : gameId,
        splitSetId : tileSetId,
        index : index,
        moveLocation : location
    }
}

function msgHandToTable(playerId, gameId, tileSetId) {
    return {
        messageType : "PlayerMove",
        moveType : "HAND_TO_TABLE",
        playerId : playerId,
        gameId : gameId,
        tileSetId : tileSetId,
    }
}

function msgTableToHand(playerId, gameId, tileSetId) {
    return {
        messageType : "PlayerMove",
        moveType : "TABLE_TO_HAND",
        playerId : playerId,
        gameId : gameId,
        tileSetId : tileSetId,
    }
}

function msgEndTurn(playerId, gameId) {
    return {
        messageType : "PlayerMove",
        moveType : "END_TURN",
        playerId : playerId,
        gameId : gameId
    }
}

function msgResetTurn(playerId, gameId) {
    return {
        messageType : "PlayerMove",
        moveType : "RESET_TURN",
        playerId : playerId,
        gameId : gameId
    }
}

function msgTakeFromHeap(playerId, gameId) {
    return {
        messageType : "PlayerMove",
        moveType : "TAKE_FROM_HEAP",
        playerId : playerId,
        gameId : gameId
    }
}

// TODO messages for PlayerMove

export {
    msgCreateGame,
    msgRemoveGame,
    msgJoinGame,
    msgLeaveGame,
    msgStartGame,
    msgStopGame,
    msgRequestGameList,
    msgRequestPlayerList,
    msgRequestPlayerListForGame,
    msgRequestGameState,
    msgMerge,
    msgMoveAndMerge,
    msgSplit,
    msgHandToTable,
    msgTableToHand,
    msgEndTurn,
    msgResetTurn,
    msgTakeFromHeap
};
